frame/babe/src/lib.rs

```diff
@@ -571,38 +580,47 @@ impl<T: Trait> Module<T> {
 			Lateness::<T>::put(lateness);
 			CurrentSlot::put(current_slot);
 
-			if let PreDigest::Primary(primary) = digest {
-				// place the VRF output into the `Initialized` storage item
-				// and it'll be put onto the under-construction randomness
-				// later, once we've decided which epoch this block is in.
-				//
-				// Reconstruct the bytes of VRFInOut using the authority id.
-				Authorities::get()
-					.get(primary.authority_index as usize)
-					.and_then(|author| {
-						schnorrkel::PublicKey::from_bytes(author.0.as_slice()).ok()
-					})
-					.and_then(|pubkey| {
-						let transcript = sp_consensus_babe::make_transcript(
-							&Self::randomness(),
-							current_slot,
-							EpochIndex::get(),
-						);
-
-						primary.vrf_output.0.attach_input_hash(
-							&pubkey,
-							transcript
-						).ok()
-					})
-					.map(|inout| {
-						inout.make_bytes(&sp_consensus_babe::BABE_VRF_INOUT_CONTEXT)
-					})
-			} else {
-				None
-			}
+			let authority_index = digest.authority_index();
+
+			// Extract out the VRF output if we have it
+			digest
+				.vrf_output()
+				.and_then(|vrf_output| {
+					// place the VRF output into the `Initialized` storage item
+					// and it'll be put onto the under-construction randomness
+					// later, once we've decided which epoch this block is in.
+					//
+					// Place the both primary and secondary VRF output into the
+					// `AuthorVrfRandomness` storage item.
+					//
+					// Reconstruct the bytes of VRFInOut using the authority id.
+					Authorities::get()
+						.get(authority_index as usize)
+						.and_then(|author| {
+							schnorrkel::PublicKey::from_bytes(author.0.as_slice()).ok()
+						})
+						.and_then(|pubkey| {
+							let transcript = sp_consensus_babe::make_transcript(
+								&Self::randomness(),
+								current_slot,
+								EpochIndex::get(),
+							);
+
+							vrf_output.0.attach_input_hash(
+								&pubkey,
+								transcript
+							).ok()
+						})
+						.map(|inout| {
+							inout.make_bytes(&sp_consensus_babe::BABE_VRF_INOUT_CONTEXT)
+						})
+			})
 		});
 
-		Initialized::put(maybe_randomness);
+		let maybe_primary_randomness = maybe_randomness.filter(|_| is_primary);
+
+		Initialized::put(maybe_primary_randomness);
```

Comment:
I would find this cleaner to read just as a single `if is_primary { Initializer:put(maybe_randomness) }`

Reply:
Note that this is not exactly the same as what the existing version does, which puts a `None` when there is no randomness

Reply:
You could do
```rust
Initialized::put(if is_primary { maybe_randomness } else { None });
```

Reply:
The base value of `Initialized` is `None`, so it logically doesn't make a difference. But yeah, the second option is more exact

Reply:
I tried the suggested approach of `if is_primary { Initialized::put(maybe_randomness) }` but got assertion failures here https://github.com/paritytech/substrate/blob/e65a60cf09f76ae6a80aab1f0653eef84a8e80db/frame/babe/src/lib.rs#L464 in a lot of the frame-babe tests